### ISSUE TITLE
Information bottleneck in E model architecture

### ISSUE THREAD

**Issue by vladkvit** [sgrvinod/chess-transformers#14]
There are roughly 2000 possible moves that the E-type models are predicting. Your `d_model` parameter is 512 for the CT-E-20 model, which is smaller than 2000. You have the following line in your model:
`moves = self.fc(boards[:, :1, :])  # (N, 1, vocab_size)`

The dimension of `boards[:, :1, :]` is `(N, 1, d_model)`, which is then expanded to `(N, 1, vocab_size)`. 

I am writing a chess engine heavily inspired by your code. I found that removing this dimensional bottleneck (i.e. doing something like `boards[:, :4, :].reshape( BATCH_SIZE, -1 )`) will greatly enhance predictive performance. In my own testing, I found that the more of the tensor you use, the better predictive performance you get, at the cost of having a giant FC at the end.

I also found some ways of reducing the parameter count of the FC while keeping most of the performance. I can write about what worked for me if there is interest.

**Comment by EaswarGn** [comment]
Hello, which model did you make these changes for and what was the exact increase in predictive performance due to your changes?

**Comment by vladkvit** [comment]
I made these changes to my own custom model. It's not on Github, but it has a pretty similar architecture to the E model - a stack of TransformerEncoder layers that predicts the next half-move. I also predict the game outcome in my model.

With the bottleneck and a half hour training run (70k game positions), my validation loss for move prediction is 3.034. Without the bottleneck, it's 1.863.
![Screenshot 2024-09-26 184513](https://github.com/user-attachments/assets/750862af-ad2a-4199-b757-d523fb24f540)

**Comment by EaswarGn** [comment]
Just curious why did you choose to make your model closely related to the [CT-E-20](https://github.com/sgrvinod/chess-transformers?tab=readme-ov-file#ct-e-20) model instead of the [CT-EFT-85](https://github.com/sgrvinod/chess-transformers?tab=readme-ov-file#ct-eft-85) because it seems like CT-EFT-85 is just a larger version of CT-E-20 and it performs slightly better.

**Comment by sgrvinod** [comment]
Thanks! I'm surprised there is a bottleneck. I'd imagine that the network would've filtered the most important information (i.e. the move to make) by that stage. I pretty much used a single vector because that's typically what is done with a transformer used for classification in NLP: using the [CLS] token or whatever it's called.

Was this the only change you made? Did you use 4 vectors? (Which vectors specifically did you use?) I'm now wondering if it's worth average-pooling or max-pooling across all the non-board positions, if there is a bottleneck. 

In the plot you posted, is this the same dataset I used? What was the top-1 validation accuracy?

**Comment by sgrvinod** [comment]
@EaswarGn The EFT models have a different structure at the prediction stage: they directly label the board positions as From or To squares. Therefore, the vectors at all board positions are already used and there wouldn't be a bottleneck (of this sort, anyway).

**Comment by vladkvit** [comment]
> In the plot you posted, is this the same dataset I used? What was the top-1 validation accuracy?

No. I'm using PGNs from Lichess's open database, 2013-07 through 2013-09 for training and 2013-10 for validation, with about half the games filtered out. It is a much smaller dataset size (less than a million games). My data loader reads PGNs instead of HDF5, so unfortunately I can't easily run my net on the datasets used here.

I re-ran the two benchmarks with top-1 accuracy. However, do note that the datasets are different, thus the absolute values aren't directly comparable. Plus, I'm running a smaller model (about half the parameters, 5 layers and 128D embeddings).

![screencapture-localhost-6006-2024-09-27-11_50_33](https://github.com/user-attachments/assets/f91ff937-7f42-461f-94a3-e15410fc5264)

However, I noticed that a lot of architecture improvements from Leela Chess Zero (such as smolgen) ported over well to my model, so I was hoping that the move prediction improvement is not too specific to the architecture but is more general to the domain.

> why did you choose to make your model closely related to the [CT-E-20](https://github.com/sgrvinod/chess-transformers?tab=readme-ov-file#ct-e-20) model instead of the [CT-EFT-85](https://github.com/sgrvinod/chess-transformers?tab=readme-ov-file#ct-eft-85)

My goal with my model is to mimic other players as opposed to being the strongest, thus I need the full move prediction vector. I want to be able to classify positions with only one human-like move vs many human-like moves.

> I'd imagine that the network would've filtered the most important information (i.e. the move to make) by that stage.

My intuition was that encoding the move should be easier when you have separate dimensions per move. With LLMs, I'm guessing that it's a mix of causal attention masks, they reuse the token_id -> embedding matrix as the final FC layer (something that doesn't apply to chess), and they train for much longer.

> Did you use 4 vectors? (Which vectors specifically did you use?)

I found that it didn't seem to matter which specific vectors I used. I found that the more of the (64 x emb_dim) tensor I use, the better the performance (at the cost of many more parameters, so it's not a parameter-controlled comparison). Pooling (max-pool, avg-pool) sounds like a promising direction.

> Was this the only change you made?

Between the two graphs, yes.

Here's the updated code for reference (it's suboptimal from a parameter point of view, but it gets the point across):

````
reduction_dim=250
self.fc_move_reduce = nn.Linear(8 * self.d_model, reduction_dim)
self.fc_final = nn.Linear(reduction_dim * 8, self.vocab_sizes["moves"])
---
num_move_heads = 8
move_heads = boards.view(BATCH_SIZE, num_move_heads, BOARD_TOKENS // num_move_heads, self.d_model) # batch, 8, 8, 128
reshaped_heads = move_heads.reshape(BATCH_SIZE, 8, -1) # batch, 8, 8*128
reduced_heads = self.fc_move_reduce(reshaped_heads) # batch, 8, 8*128
combined = reduced_heads.reshape(BATCH_SIZE, -1)
move_preds = self.fc_final(combined)
```
